taskmanager/taskq.go

```diff
@@ -89,76 +96,88 @@ func (c *Client) loadTaskQ() error {
+		if panicErr := recover(); panicErr != nil {
+			err = fmt.Errorf(fmt.Sprintf("registering task panic: %v", panicErr))
 		}
 	}()
 
 	mutex.Lock()
+	defer mutex.Unlock()
 
-	// Check if task is already registered
-	if t := taskq.Tasks.Get(task.Name); t != nil {
-
-		// Register the task locally
-		c.options.taskq.tasks[task.Name] = t
+	if t := taskq.Tasks.Get(name); t != nil {
+		// if already registered - register the task locally
+		c.options.taskq.tasks[name] = t
+	} else {
+		// Register and store the task
+		c.options.taskq.tasks[name] = taskq.RegisterTask(&taskq.TaskOptions{
+			Name:       name,
+			Handler:    handler,
+			RetryLimit: 1,
+		})
+	}
 
-		// Task already exists!
-		// c.DebugLog(fmt.Sprintf("registering task: %s... task already exists!", task.Name))
+	c.DebugLog(fmt.Sprintf("registering task: %s...", c.options.taskq.tasks[name].Name()))
+	return nil
+}
 
-		mutex.Unlock()
+// RunTask will run a task using TaskQ
+func (c *TaskManager) RunTask(ctx context.Context, options *TaskRunOptions) error {
+	c.options.logger.Info().Msgf("executing task: %s", options.TaskName)
 
-		return
+	// Try to get the task
+	task, ok := c.options.taskq.tasks[options.TaskName]
+	if !ok {
+		return fmt.Errorf("task %s not registered", options.TaskName)
 	}
 
-	// Register and store the task
-	c.options.taskq.tasks[task.Name] = taskq.RegisterTask(convertTaskToTaskQ(task))
+	// Task message will be used to add to the queue
+	taskMessage := task.WithArgs(ctx, options.Arguments...)
 
-	mutex.Unlock()
+	// There are two ways to run a task:
+	// Option 1: Run the task immediately
+	if options.RunEveryPeriod == 0 {
+		return c.options.taskq.queue.Add(taskMessage)
+	}
 
-	// Debugging
-	c.DebugLog(fmt.Sprintf("registering task: %s...", c.options.taskq.tasks[task.Name].Name()))
+	// Option 2: Run the task periodically using cron
+	// Note: The first run will be after the period has passed
+	return c.scheduleTaskWithCron(ctx, task, taskMessage, options.RunEveryPeriod)
 }
 
-// runTaskUsingTaskQ will run a task using TaskQ
-func (c *Client) runTaskUsingTaskQ(ctx context.Context, options *TaskOptions) error {
+func (c *TaskManager) scheduleTaskWithCron(ctx context.Context, task *taskq.Task, taskMessage *taskq.Message, runEveryPeriod time.Duration) error {
+	// When using Redis, we need to use a distributed timed lock to prevent the addition of the same task to the queue by multiple instances.
+	// With this approach, only one instance will add the task to the queue within a given period.
+	var tryLock func() bool
+	if c.Factory() == FactoryRedis {
+		key := fmt.Sprintf("taskq_cronlock_%s", task.Name())
 
-	// Starting the execution of the task
-	c.DebugLog(fmt.Sprintf(
-		"executing task: %s... delay: %s arguments: %s",
-		options.TaskName,
-		options.Delay.String(),
-		fmt.Sprintf("%+v", options.Arguments),
-	))
+		// The runEveryPeriod should be greater than 1 second
+		if runEveryPeriod < 1*time.Second {
+			runEveryPeriod = 1 * time.Second
```

Comment:
maybe instead of default we could return an error here

Reply:
Done.